sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala

```diff
@@ -0,0 +1,393 @@
+   * @param fn          the work to perform. Takes T and produces an output K
+   * @tparam T element type that must be AutoCloseable
+   * @tparam K `fn` result type
+   * @return a single item of type K
+   */
+  def withRetryNoSplit[T <: AutoCloseable, K](
+      input: Seq[T])
+      (fn: Seq[T] => K): K = {
+    val wrapped = AutoCloseableSeqInternal(input)
+    drainSingleWithVerification(
+      new RmmRapidsRetryAutoCloseableIterator(
+        SingleItemAutoCloseableIteratorInternal(wrapped),
+        fn))
+  }
+
+  /**
+   * Helper method to drain an iterator and ensuring that it was non-empty
+   * and it had a single item in it.
+   */
+  private def drainSingleWithVerification[K](it: Iterator[K]): K = {
+    require(it.hasNext,
+      "Couldn't drain a single item with a closed iterator!")
+    val item = it.next()
+    require(!it.hasNext,
+      "Multiple items found in the source iterator but one expected!")
+    item
+  }
+
+  /**
+   * AutoCloseable wrapper on Seq[T], returning a Seq[T] that can be closed.
+   * @param ts the Seq to wrap
+   * @tparam T the type of the items in `ts`
+   */
+  private case class AutoCloseableSeqInternal[T <: AutoCloseable](ts: Seq[T])
+      extends Seq[T] with AutoCloseable{
+    override def close(): Unit = {
+      ts.foreach(_.safeClose())
+    }
+
+    override def length: Int = ts.length
+
+    override def iterator: Iterator[T] = ts.iterator
+
+    override def apply(idx: Int): T = ts.apply(idx)
+  }
+
+  /**
+   * An iterator of a single item that is able to close if .next
+   * has not been called on it.
+   * @param ts the AutoCloseable item to close if this iterator hasn't been drained
+   * @tparam T the type of `ts`, must be AutoCloseable
+   */
+  private case class SingleItemAutoCloseableIteratorInternal[T <: AutoCloseable](ts: T)
+    extends Iterator[T] with AutoCloseable {
+
+    private var wasCalled = false
+    override def hasNext: Boolean = !wasCalled
+    override def next(): T = {
+      wasCalled = true
+      ts
+    }
+    override def close(): Unit = {
+      if (!wasCalled) {
+        ts.close()
+      }
+    }
+  }
+
+  /**
+   * RmmRapidsRetryAutoCloseableIterator exposes an iterator that can retry work,
+   * specified by `fn`, abstracting away the retry specifics. Elements passed to this iterator
+   * must be AutoCloseable.
+   *
+   * It assumes the type K is AutoCloseable, and that if a split policy is specified, that it
+   * is capable of handling splitting one K into a sequence of them.
+   *
+   * When an attempt to invoke function `fn` is successful, the item K in `input` will be
+   * closed. In the case of a failure, all attempts will be closed. It is the responsibility
+   * of the caller to close any remaining items in `input` that have not been attempted.
+   *
+   * @tparam K element type that must be AutoCloseable
+   * @tparam T `fn` result type
+   * @param input an iterator of K
+   * @param fn a function that takes K and produces T
```

Comment:
nit: Can we have something here about how `fn` can have no real side effects. Metrics and logging are okay, but the results of `fn` cannot change based off of a previous call to `fn`. And this needs to be everywhere that `fn` is in a public API.